File: commitguard/github_analyzer.py
"""GitHub API integration for remote repository analysis."""

import json
import logging
import re
from urllib.request import Request, urlopen
from urllib.error import HTTPError, URLError

from .analyzer import (
    _call_ai,
    _validate_diff_size,
    AnalysisError,
    AIAnalysisError,
    DiffTooLargeError,
)

GITHUB_API_BASE = "https://api.github.com"
logger = logging.getLogger(__name__)


class GitHubError(AnalysisError):
    """Raised when GitHub API operations fail."""


def is_github_url(url: str) -> bool:
    """Return True if the string looks like a GitHub repository URL."""
    if not url:
        return False
    url = url.strip()
    return bool(
        re.match(r"^https?://github\.com/[\w.\-]+/[\w.\-]+", url)
        or re.match(r"^git@github\.com:[\w.\-]+/[\w.\-]+", url)
    )


def parse_github_url(url: str) -> tuple[str, str]:
    """
    Extract (owner, repo) from a GitHub URL.
    Raises ValueError if the URL cannot be parsed.
    """
    url = url.strip().rstrip("/")
    m = re.match(r"^https?://github\.com/([\w.\-]+)/([\w.\-]+?)(?:\.git)?$", url)
    if m:
        return m.group(1), m.group(2)
    m = re.match(r"^git@github\.com:([\w.\-]+)/([\w.\-]+?)(?:\.git)?$", url)
    if m:
        return m.group(1), m.group(2)
    raise ValueError(f"Could not parse GitHub URL: {url!r}")


def _github_request(
    path: str,
    token: str | None = None,
    *,
    accept: str = "application/vnd.github+json",
) -> bytes:
    """Make a GET request to the GitHub API and return the response body."""
    url = f"{GITHUB_API_BASE}{path}"
    headers = {
        "Accept": accept,
        "X-GitHub-Api-Version": "2022-11-28",
        "User-Agent": "CommitGuard/1.0",
    }
    if token:
        headers["Authorization"] = f"Bearer {token}"
    req = Request(url, headers=headers)
    try:
        with urlopen(req, timeout=30) as resp:
            return resp.read()
    except HTTPError as e:
        body = e.read().decode("utf-8", errors="replace") if e.fp else ""
        try:
            data = json.loads(body)
            msg = data.get("message", body) or str(e)
        except Exception:
            msg = body or str(e)
        status = e.code
        if status == 401:
            raise GitHubError(f"GitHub authentication failed: {msg}") from e
        if status == 403:
            raise GitHubError(
                f"GitHub access forbidden (rate limit or private repo): {msg}"
            ) from e
        if status == 404:
            raise GitHubError(f"GitHub resource not found: {msg}") from e
        raise GitHubError(f"GitHub API error {status}: {msg}") from e
    except URLError as e:
        raise GitHubError(f"Could not reach GitHub API: {e}") from e


def list_github_commits(
    owner: str,
    repo: str,
    *,
    token: str | None = None,
    search: str = "",
    limit: int = 80,
    branch: str = "",
) -> list[dict[str, str]]:
    """List recent commits from a GitHub repository."""
    safe_limit = min(max(1, limit), 100)
    params = f"per_page={safe_limit}"
    if branch:
        params += f"&sha={branch}"
    body = _github_request(f"/repos/{owner}/{repo}/commits?{params}", token)
    commits_data = json.loads(body)

    term = search.strip().lower()
    commits = []
    for item in commits_data:
        sha = item.get("sha", "")
        commit_info = item.get("commit", {})
        message = (commit_info.get("message", "") or "").strip()
        title = message.split("\n")[0] if message else sha[:8]
        author_info = commit_info.get("author", {}) or {}
        author_name = (author_info.get("name", "") or "").strip() or "Unknown author"
        committed_at = author_info.get("date", "")

        if term:
            haystack = f"{sha}\n{title}\n{author_name}\n{message}".lower()
            if term not in haystack:
                continue

        commits.append(
            {
                "ref": sha,
                "short_ref": sha[:8] if sha else "commit",
                "title": title or sha[:8],
                "author": author_name,
                "date": committed_at,
            }
        )

        if len(commits) >= safe_limit:
            break

    return commits


def fetch_github_commit_diff(
    owner: str,
    repo: str,
    sha: str,
    *,
    token: str | None = None,
) -> tuple[str, str, list[str]]:
    """
    Fetch a commit's diff, message, and changed files from GitHub.
    Returns (diff_text, commit_message, files).
    """
    body = _github_request(f"/repos/{owner}/{repo}/commits/{sha}", token)
    data = json.loads(body)
    commit_info = data.get("commit", {})
    message = (commit_info.get("message", "") or "").strip() or sha[:8]
    files_data = data.get("files", [])
    file_paths = [f.get("filename", "") for f in files_data if f.get("filename")]

    diff_body = _github_request(
        f"/repos/{owner}/{repo}/commits/{sha}",
        token,
        accept="application/vnd.github.v3.diff",
    )
    diff_text = diff_body.decode("utf-8", errors="replace")
    return diff_text, message, file_paths


def analyze_github_commit(
    owner: str,
    repo: str,
    sha: str,
    *,
    token: str | None = None,
    api_key: str,
    model: str = "openai/gpt-4o-mini",
    max_diff_chars=None,
    system_prompt: str | None = None,
) -> tuple[str, str]:
    """Analyze a GitHub commit. Returns (analysis_result, diff)."""
    try:
        diff, message, files = fetch_github_commit_diff(owner, repo, sha, token=token)
    except GitHubError:
        raise
    except Exception as e:
        raise GitHubError(f"Could not fetch commit {sha[:8]!r}: {e}") from e

    _validate_diff_size(diff, context=f"commit {sha[:8]}", max_diff_chars=max_diff_chars)

    try:
        result = _call_ai(
            diff, message, files, api_key, model, system_prompt=system_prompt
        )
    except DiffTooLargeError:
        raise
    except Exception as e:
        raise AIAnalysisError(f"AI analysis failed: {e}") from e

    return result, diff


def analyze_github_commit_range(
    owner: str,
    repo: str,
    base: str,
    head: str,
    *,
    token: str | None = None,
    api_key: str,
    model: str = "openai/gpt-4o-mini",
    max_commits: int = 20,
    max_diff_chars=None,
    system_prompt: str | None = None,
) -> list[dict[str, str]]:
    """
    Analyze commits in a range on GitHub (base..head).
    Returns list of analysis results, newest-first.
    """
    try:
        body = _github_request(
            f"/repos/{owner}/{repo}/compare/{base}...{head}", token
        )
        data = json.loads(body)
        commits = data.get("commits", [])
        if not commits:
            raise GitHubError(f"No commits found between {base!r} and {head!r}")
        commits = list(reversed(commits[:max_commits]))
    except GitHubError:
        raise
    except Exception as e:
        raise GitHubError(f"Could not fetch commit range: {e}") from e

    analyses = []
    for item in commits:
        sha = item.get("sha", "")
        commit_info = item.get("commit", {})
        title = (commit_info.get("message", "") or "").split("\n")[0] or sha[:8]
        try:
            result, diff = analyze_github_commit(
                owner,
                repo,
                sha,
                token=token,
                api_key=api_key,
                model=model,
                max_diff_chars=max_diff_chars,
                system_prompt=system_prompt,
            )
            analyses.append(
                {
                    "ref": sha,
                    "short_ref": sha[:8],
                    "title": title,
                    "result": result,
                    "diff": diff,
                }
            )
        except DiffTooLargeError as e:
            raise DiffTooLargeError(
                f"Commit {sha[:8]} cannot be analyzed: {e}"
            ) from e
        except Exception as e:
            raise AIAnalysisError(
                f"AI analysis failed for commit {sha[:8]}: {e}"
            ) from e

    return analyses


def list_github_prs(
    owner: str,
    repo: str,
    *,
    token: str | None = None,
    state: str = "open",
    limit: int = 50,
) -> list[dict]:
    """List pull requests from a GitHub repository."""
    safe_limit = min(max(1, limit), 100)
    body = _github_request(
        f"/repos/{owner}/{repo}/pulls?state={state}&per_page={safe_limit}&sort=updated&direction=desc",
        token,
    )
    prs_data = json.loads(body)
    return [
        {
            "number": pr.get("number"),
            "title": (pr.get("title") or "").strip(),
            "author": (pr.get("user", {}) or {}).get("login", ""),
            "state": pr.get("state", ""),
            "draft": pr.get("draft", False),
            "base": (pr.get("base", {}) or {}).get("ref", ""),
            "head": (pr.get("head", {}) or {}).get("ref", ""),
            "created_at": pr.get("created_at", ""),
            "updated_at": pr.get("updated_at", ""),
            "url": pr.get("html_url", ""),
        }
        for pr in prs_data
    ]


def analyze_github_pr(
    owner: str,
    repo: str,
    pr_number: int,
    *,
    token: str | None = None,
    api_key: str,
    model: str = "openai/gpt-4o-mini",
    max_diff_chars=None,
    system_prompt: str | None = None,
) -> tuple[str, str, str]:
    """
    Analyze a GitHub pull request. Returns (analysis_result, diff, pr_title).
    """
    try:
        body = _github_request(f"/repos/{owner}/{repo}/pulls/{pr_number}", token)
        pr_data = json.loads(body)
        pr_title = (pr_data.get("title") or "").strip() or f"PR #{pr_number}"
    except GitHubError:
        raise
    except Exception as e:
        raise GitHubError(f"Could not fetch PR #{pr_number}: {e}") from e

    try:
        diff_body = _github_request(
            f"/repos/{owner}/{repo}/pulls/{pr_number}",
            token,
            accept="application/vnd.github.v3.diff",
        )
        diff = diff_body.decode("utf-8", errors="replace")
    except GitHubError:
        raise
    except Exception as e:
        raise GitHubError(f"Could not fetch diff for PR #{pr_number}: {e}") from e

    try:
        files_body = _github_request(
            f"/repos/{owner}/{repo}/pulls/{pr_number}/files?per_page=100", token
        )
        files_data = json.loads(files_body)
        files = [f.get("filename", "") for f in files_data if f.get("filename")]
    except Exception:
        files = []

    _validate_diff_size(diff, context=f"PR #{pr_number}", max_diff_chars=max_diff_chars)

    try:
        result = _call_ai(
            diff,
            f"PR #{pr_number}: {pr_title}",
            files,
            api_key,
            model,
            system_prompt=system_prompt,
        )
    except DiffTooLargeError:
        raise
    except Exception as e:
        raise AIAnalysisError(f"AI analysis failed: {e}") from e

    return result, diff, pr_title

File: webui/src/lib/github-analyzer.ts
const GITHUB_API_BASE = "https://api.github.com";

export class GitHubError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "GitHubError";
  }
}

export function isGithubUrl(url: string): boolean {
  if (!url?.trim()) return false;
  const u = url.trim();
  return /^https?:\/\/github\.com\/[\w.\-]+\/[\w.\-]+/.test(u) || /^git@github\.com:[\w.\-]+\/[\w.\-]+/.test(u);
}

export function parseGithubUrl(url: string): [string, string] {
  const u = url.trim().replace(/\/$/, "");
  let m = u.match(/^https?:\/\/github\.com\/([\w.\-]+)\/([\w.\-]+?)(?:\.git)?$/);
  if (m) return [m[1], m[2]];
  m = u.match(/^git@github\.com:([\w.\-]+)\/([\w.\-]+?)(?:\.git)?$/);
  if (m) return [m[1], m[2]];
  throw new Error(`Could not parse GitHub URL: ${url}`);
}

async function githubRequest(
  path: string,
  token?: string | null,
  accept = "application/vnd.github+json"
): Promise<Response> {
  const res = await fetch(`${GITHUB_API_BASE}${path}`, {
    headers: {
      Accept: accept,
      "X-GitHub-Api-Version": "2022-11-28",
      "User-Agent": "CommitGuard/1.0",
      ...(token && { Authorization: `Bearer ${token}` }),
    },
  });

  if (!res.ok) {
    const body = await res.text();
    let msg = body;
    try {
      const data = JSON.parse(body);
      msg = data.message || body;
    } catch {}
    if (res.status === 401) throw new GitHubError(`GitHub authentication failed: ${msg}`);
    if (res.status === 403) throw new GitHubError(`GitHub access forbidden (rate limit or private repo): ${msg}`);
    if (res.status === 404) throw new GitHubError(`GitHub resource not found: ${msg}`);
    throw new GitHubError(`GitHub API error ${res.status}: ${msg}`);
  }
  return res;
}

export async function listGithubCommits(
  owner: string,
  repo: string,
  token?: string | null,
  search = "",
  limit = 80
): Promise<Array<{ ref: string; short_ref: string; title: string; author: string; date: string }>> {
  const safeLimit = Math.min(Math.max(1, limit), 100);
  const res = await githubRequest(`/repos/${owner}/${repo}/commits?per_page=${safeLimit}`, token);
  const commitsData = (await res.json()) as Array<{
    sha: string;
    commit: { message?: string; author?: { name?: string; date?: string } };
  }>;

  const term = search.trim().toLowerCase();
  const commits: Array<{ ref: string; short_ref: string; title: string; author: string; date: string }> = [];

  for (const item of commitsData) {
    const sha = item.sha || "";
    const message = (item.commit?.message || "").trim();
    const title = message ? message.split("\n")[0] : sha.slice(0, 8);
    const authorName = (item.commit?.author?.name || "").trim() || "Unknown author";
    const committedAt = item.commit?.author?.date || "";

    if (term) {
      const haystack = `${sha}\n${title}\n${authorName}\n${message}`.toLowerCase();
      if (!haystack.includes(term)) continue;
    }

    commits.push({
      ref: sha,
      short_ref: sha.slice(0, 8),
      title: title || sha.slice(0, 8),
      author: authorName,
      date: committedAt,
    });
    if (commits.length >= safeLimit) break;
  }
  return commits;
}

export async function fetchGithubCommitDiff(
  owner: string,
  repo: string,
  sha: string,
  token?: string | null
): Promise<[string, string, string[]]> {
  const res = await githubRequest(`/repos/${owner}/${repo}/commits/${sha}`, token);
  const data = (await res.json()) as { commit?: { message?: string }; files?: Array<{ filename?: string }> };
  const message = (data.commit?.message || "").trim() || sha.slice(0, 8);
  const files = (data.files || []).map((f) => f.filename || "").filter(Boolean);

  const diffRes = await githubRequest(
    `/repos/${owner}/${repo}/commits/${sha}`,
    token,
    "application/vnd.github.v3.diff"
  );
  const diffText = await diffRes.text();
  return [diffText, message, files];
}

export async function analyzeGithubCommit(
  owner: string,
  repo: string,
  sha: string,
  apiKey: string,
  model: string,
  token?: string | null,
  maxDiffChars?: number | string | null,
  systemPrompt?: string | null
): Promise<{ result: string; diff: string; short_ref: string; title: string; author: string; date: string }> {
  const { callAi, validateDiffSize } = await import("./analyzer");
  const [diff, message, files] = await fetchGithubCommitDiff(owner, repo, sha, token);
  validateDiffSize(diff, `commit ${sha.slice(0, 8)}`, maxDiffChars);
  const result = await callAi(diff, message, files, apiKey, model, systemPrompt);
  const title = message.split("\n")[0] || sha.slice(0, 8);
  return { result, diff, short_ref: sha.slice(0, 8), title, author: "", date: "" };
}

export async function analyzeGithubCommitRange(
  owner: string,
  repo: string,
  base: string,
  head: string,
  apiKey: string,
  model: string,
  token?: string | null,
  maxCommits = 20,
  maxDiffChars?: number | string | null,
  systemPrompt?: string | null
): Promise<Array<{ ref: string; short_ref: string; title: string; author: string; date: string; result: string; diff: string }>> {
  const { callAi, AIAnalysisError, DiffTooLargeError } = await import("./analyzer");
  const res = await githubRequest(`/repos/${owner}/${repo}/compare/${base}...${head}`, token);
  const data = (await res.json()) as { commits?: Array<{ sha?: string; commit?: { message?: string; author?: { date?: string; name?: string } }; author?: { login?: string } }> };
  const commits = (data.commits || []).slice(0, maxCommits).reverse();
  if (!commits.length) throw new GitHubError(`No commits found between '${base}' and '${head}'`);

  const analyses: Array<{ ref: string; short_ref: string; title: string; author: string; date: string; result: string; diff: string }> = [];
  for (const item of commits) {
    const sha = item.sha || "";
    const title = (item.commit?.message || "").split("\n")[0] || sha.slice(0, 8);
    const author = item.commit?.author?.name || item.author?.login || "";
    const date = item.commit?.author?.date || "";
    try {
      const { result, diff } = await analyzeGithubCommit(
        owner,
        repo,
        sha,
        apiKey,
        model,
        token,
        maxDiffChars,
        systemPrompt
      );
      analyses.push({ ref: sha, short_ref: sha.slice(0, 8), title, author, date, result, diff });
    } catch (e) {
      if (e instanceof DiffTooLargeError) {
        throw new DiffTooLargeError(`Commit ${sha.slice(0, 8)} cannot be analyzed: ${e}`);
      }
      throw new AIAnalysisError(`AI analysis failed for commit ${sha.slice(0, 8)}: ${e}`);
    }
  }
  return analyses;
}

export async function analyzeGithubPr(
  owner: string,
  repo: string,
  prNumber: number,
  apiKey: string,
  model: string,
  token?: string | null,
  maxDiffChars?: number | string | null,
  systemPrompt?: string | null
): Promise<{ result: string; diff: string; prTitle: string }> {
  const { callAi, validateDiffSize } = await import("./analyzer");
  const res = await githubRequest(`/repos/${owner}/${repo}/pulls/${prNumber}`, token);
  const prData = (await res.json()) as { title?: string };
  const prTitle = (prData.title || "").trim() || `PR #${prNumber}`;

  const diffRes = await githubRequest(
    `/repos/${owner}/${repo}/pulls/${prNumber}`,
    token,
    "application/vnd.github.v3.diff"
  );
  const diff = await diffRes.text();

  let files: string[] = [];
  try {
    const filesRes = await githubRequest(
      `/repos/${owner}/${repo}/pulls/${prNumber}/files?per_page=100`,
      token
    );
    const filesData = (await filesRes.json()) as Array<{ filename?: string }>;
    files = filesData.map((f) => f.filename || "").filter(Boolean);
  } catch {
    // ignore
  }

  validateDiffSize(diff, `PR #${prNumber}`, maxDiffChars);
  const result = await callAi(diff, `PR #${prNumber}: ${prTitle}`, files, apiKey, model, systemPrompt);
  return { result, diff, prTitle };
}

export async function listGithubPrs(
  owner: string,
  repo: string,
  token?: string | null,
  state = "open",
  limit = 50
): Promise<
  Array<{
    number: number;
    title: string;
    author: string;
    state: string;
    draft: boolean;
    base: string;
    head: string;
    updated_at: string;
  }>
> {
  const safeLimit = Math.min(Math.max(1, limit), 100);
  const res = await githubRequest(
    `/repos/${owner}/${repo}/pulls?state=${state}&per_page=${safeLimit}&sort=updated&direction=desc`,
    token
  );
  const prsData = (await res.json()) as Array<{
    number: number;
    title?: string;
    user?: { login?: string };
    state?: string;
    draft?: boolean;
    base?: { ref?: string };
    head?: { ref?: string };
    updated_at?: string;
  }>;

  return prsData.map((pr) => ({
    number: pr.number,
    title: (pr.title || "").trim(),
    author: (pr.user?.login || "").trim(),
    state: pr.state || "",
    draft: pr.draft || false,
    base: (pr.base?.ref || "").trim(),
    head: (pr.head?.ref || "").trim(),
    updated_at: pr.updated_at || "",
  }));
}
